### ISSUE TITLE
Have @org.jspecify.annotations.Nullable on all the fields of the Generated Builder 

### ISSUE THREAD

**Issue by alexandrenavarro** [skinny85/jilt#27]
Have `@org.jspecify.annotations.Nullable` on all the fields of the Generated Builder 

I used Jilt staged builder with nullaway annotation processor to have a maximum of  potential NPE errors detected at build time. Today, because of not having `@org.jspecify.annotations.Nullable` on all the fields of the generated builder, I must add the option of `-XepOpt:NullAway:TreatGeneratedAsUnannotated` in nullaway.

But the problem is it removes all the nullaway checks on the calls of the generated builder class. 
Typically I can call `Person.builder().firstName(null).lastName(null).build()` in my code even if the fields  `firstName` / `lastName` are mandatory without any compiler errors which is problematic in my point of view. 
If all the fields are annotated with `@org.jspecify.annotations.Nullable` and I can remove this option -`XepOpt:NullAway:TreatGeneratedAsUnannotated` and nullaway will generate compilation errors for this use case.

A potential solution is to let the user to configure extra annotations in the `@Builder` annotation on the fields on the generated builder class, like `"extraAnnotationsOnBuilderFields" = "@org.jspecify.annotations.Nullable"`
in order not to to force user to use jspecify if the do not want (compared if you set it directly in the generated code which is also a solution).
In the future, it will be great to have fields with `?`  in the builder when this jep https://openjdk.org/jeps/8316779 will be released   This jep manages a way to indicate in a field is nullable or not directly in the java language.

Thank you for your useful lib and don't hesitate if you have questions.

**Comment by skinny85** [comment]
Thanks for opening the issue @alexandrenavarro.

Can you please set up a simple example project that exhibits this issue, and share it here as a public GitHub repo (kind of like what was done [here](https://github.com/skinny85/jilt/issues/14#issuecomment-2018258861))? Like we discussed this before, I think you should turn off static analysis for generated code, and code like `Person.builder().firstName(null)` will still be covered by these checks, as it's not generated. So, I want to make sure I understand correctly what the issue is here.

Thanks,
Adam

**Comment by alexandrenavarro** [comment]
Ok I will do, I will keep in touch.

**Comment by alexandrenavarro** [comment]
Hi,

I made a sample to show you the problem https://github.com/alexandrenavarro/jilt-problem-sample and I discovered something a bit particular on nullaway.

If you use
 -XepOpt:NullAway:TreatGeneratedAsUnannotated
 on your generated classes, nullaway will not check potential NPE on external and internal API.

If you use
-XepOpt:NullAway:ExcludedClassAnnotations=javax.annotation.processing.Generated
on your generated classes, nullaway will check potential NPE on external API but not on internal API.
I knew/tested only the first use case when I created the issue.

As jilt Builder are well annotated on external API (but not on internal API), using 
-XepOpt:NullAway:ExcludedClassAnnotations=javax.annotation.processing.Generated
 can be a workaround.
Nevertheless if you use many generators which can use the same @Generated class, you can not use this workaround if you want to use different settings if some generated classes are well annotated on internal or/and external API and some are not.

Solution 1 :
Having an option to specify some extra annotations on the builder class (like x-class-extra-annotation on openapi-generator for instance) or having a specific jilt @Generated class can solve the problem.
It also solved this issue https://github.com/skinny85/jilt/issues/28.

Solution 2: 
Another way to fix the problem is to generate a correct builder for the nullaway checker.
To do it it, you must :
- 1 add @jspecify.Nullable on each field of the builder
- 2 check in the build() method if all mandatory fields are really not null before creating the instance or throw an exception like IllegalArgumentException.
The throw of an exception should not occur when you use StageBuilder and nullaway at the same time but it can occur if you do not use one of them. You can imagine implementing it only on StageBuilder.

The first point can be solved by adding an extra annotation on the field (like x-field-extra-annotation on openapi-generator) or force add @jspecify.Nullable but it forces the client to have jspecify as dependency (personally, I'm ok but I can understand it is problematic).

The second point can be fixed
- in the generator just for StagedBuilder for instance,
- through a private constructor on @ Builder and implement the build method (possible but painful to do it on each class).

Don't hesitate if it is not clear, if you have some questions or any other potential solutions to solve this issue.

In my opinion : 
The solution 1 is pragmatic, is not so complicated and solves the problem from a client point of view.
The solution 2 is nicer / cleaner but needs to think how to do it correctly and it is not so simple to do correctly, many things 
 to checked to have a perfect solution.

Thank you in advance,

**Comment by skinny85** [comment]
Thanks for the update @alexandrenavarro.

I want to zoom into this part:

> As jilt Builder are well annotated on external API (but not on internal API), using `-XepOpt:NullAway:ExcludedClassAnnotations=javax.annotation.processing.Generated` can be a workaround.
> Nevertheless if you use many generators which can use the same `@Generated` class, you can not use this workaround if you want to use different settings if some generated classes are well annotated on internal or/and external API and some are not.

This is the part that I never understood.

Why would you _ever_ want to run static analysis like ErrorProne / NullAway on generated code?

Even if it finds a problem -  you won't be able to fix it anyway, because the code is generated, and you don't control the generator.

So what's the point of running that analysis at all?

In my opinion, you should **always** run ErrorProne / NullAway with `-XepOpt:NullAway:ExcludedClassAnnotations=javax.annotation.processing.Generated`, or exclude the generated code with the `-XepExcludedPaths` option, like they explain in [their documentation](https://github.com/uber/NullAway?tab=readme-ov-file#annotation-processors--generated-code).

If you do that, I think the problem naturally goes away, without having to make any changes to Jilt, or your build's configuration.

Let me know if that makes sense.

**Comment by skinny85** [comment]
> Because of the options of nullaway on the exclusions of checks on external api, I think I need more to have a custom annotation on generated code not well annotated in practice than a custom annotation on generated code well annotated like jilt.

Yes, if the generated code doesn't correctly annotate its public API for `null` checking, you won't be able to run NullAway on hand-written code using that generated code. However, no Jilt feature will help you fix that problem - in that case, you either need to exclude that hand-written code from NullAway analysis, or change the generator you use.

**Comment by alexandrenavarro** [comment]
I did some tests on more complex and I confirmed, it will not help to have a custom generated annotation on jilt because of the way you can configure nullaway notably the exclusion of some classes.
Sorry for the time you spent on this issue. I think you can close the issue.

**Comment by skinny85** [comment]
No worries, I really didn't spend that much time on it, and the project you provided as a reproduction is super valuable, so thanks a lot! 🙂

Feel free to comment if you ever want the issue re-opened.
